File: python/src/graph_engineering/builder.py
"""Detached, deterministic general Graph IR builder."""

from __future__ import annotations

import json
from dataclasses import dataclass
from enum import StrEnum
from typing import Final, TypeVar, cast

from pydantic import BaseModel, ValidationError

from .compiler import Diagnostic, try_compile_graph
from .component_identity import (
    CompiledGraphIdentity,
    _create_compiled_graph_identity_from_compiled,
)
from .models import (
    EdgeSpec,
    Endpoint,
    GraphModelSnapshotError,
    GraphPolicies,
    GraphSpec,
    JsonObject,
    Metadata,
    NodeSpec,
    capture_graph_model_document,
)
from .portable_json import PortableJsonError, portable_json_snapshot
from .typed_ports import TYPED_PORT_POLICY_KEY, strict_typed_ports_policy

_MISSING: Final = object()
ModelT = TypeVar("ModelT", bound=BaseModel)


class BuilderErrorCode(StrEnum):
    INVALID_INPUT = "GE_BUILDER_INVALID_INPUT"
    DUPLICATE_NODE = "GE_BUILDER_DUPLICATE_NODE"
    DUPLICATE_EDGE = "GE_BUILDER_DUPLICATE_EDGE"
    DUPLICATE_ENTRYPOINT = "GE_BUILDER_DUPLICATE_ENTRYPOINT"
    DUPLICATE_OUTPUT = "GE_BUILDER_DUPLICATE_OUTPUT"
    MISSING_REQUIRED = "GE_BUILDER_MISSING_REQUIRED"
    CORE_REJECTED = "GE_BUILDER_CORE_REJECTED"
    SEALED = "GE_BUILDER_SEALED"


def _diagnostic_dict(diagnostic: Diagnostic) -> JsonObject:
    return {
        "code": diagnostic.code.value,
        "message": diagnostic.message,
        "nodeId": diagnostic.node_id,
        "nodeIds": list(diagnostic.node_ids) if diagnostic.node_ids is not None else None,
        "edgeId": diagnostic.edge_id,
        "outputName": diagnostic.output_name,
        "path": diagnostic.path,
    }


class GraphBuilderError(ValueError):
    """A stable builder failure with optional canonical compiler diagnostics."""

    def __init__(
        self,
        code: BuilderErrorCode,
        message: str,
        path: str,
        diagnostics: tuple[Diagnostic, ...] = (),
    ) -> None:
        self.code = code
        self.path = path
        self.diagnostics = diagnostics
        super().__init__(message)

    @property
    def message(self) -> str:
        return str(self)

    def to_dict(self) -> JsonObject:
        return {
            "code": self.code.value,
            "message": self.message,
            "path": self.path,
            "diagnostics": [_diagnostic_dict(item) for item in self.diagnostics],
        }


def _fail(
    code: BuilderErrorCode,
    message: str,
    path: str,
    diagnostics: tuple[Diagnostic, ...] = (),
) -> GraphBuilderError:
    return GraphBuilderError(code, message, path, diagnostics)


def _model_snapshot(
    value: object,
    model_type: type[ModelT],
    path: str,
) -> JsonObject:
    try:
        document: object
        if type(value) is model_type:
            document = capture_graph_model_document(value, model_type)
        elif type(value) is dict:
            document = portable_json_snapshot(value)
        else:
            raise PortableJsonError("model input must be a plain object")
        validated = model_type.model_validate(document)
        snapshot = portable_json_snapshot(capture_graph_model_document(validated, model_type))
    except (
        GraphModelSnapshotError,
        PortableJsonError,
        ValidationError,
        RecursionError,
        TypeError,
        ValueError,
    ):
        raise _fail(
            BuilderErrorCode.INVALID_INPUT,
            "builder input is not a valid detached Graph IR object",
            path,
        ) from None
    if not isinstance(snapshot, dict):
        raise _fail(
            BuilderErrorCode.INVALID_INPUT,
            "builder input must be a Graph IR object",
            path,
        )
    return snapshot


def _json_object_snapshot(value: object, path: str) -> JsonObject:
    try:
        if type(value) is not dict:
            raise PortableJsonError("value must be a plain object")
        snapshot = portable_json_snapshot(value)
    except (PortableJsonError, RecursionError):
        raise _fail(
            BuilderErrorCode.INVALID_INPUT,
            "builder value must be a detached portable JSON object",
            path,
        ) from None
    if not isinstance(snapshot, dict):
        raise _fail(
            BuilderErrorCode.INVALID_INPUT,
            "builder value must be a JSON object",
            path,
        )
    return snapshot


def _string_snapshot(value: object, path: str, *, non_empty: bool) -> str:
    try:
        snapshot = portable_json_snapshot(value)
    except PortableJsonError:
        raise _fail(
            BuilderErrorCode.INVALID_INPUT,
            "builder value must be a portable string",
            path,
        ) from None
    if not isinstance(snapshot, str) or (non_empty and not snapshot):
        message = (
            "builder value must be a non-empty string"
            if non_empty
            else "builder value must be a string"
        )
        raise _fail(
            BuilderErrorCode.INVALID_INPUT,
            message,
            path,
        )
    return snapshot


@dataclass(frozen=True, slots=True)
class BuiltGraph:
    """Immutable identity fields plus a detached-on-access GraphSpec snapshot."""

    _canonical_graph: str
    graph_hash: str
    identity: CompiledGraphIdentity

    @property
    def canonical_graph(self) -> str:
        return self._canonical_graph

    @property
    def graph(self) -> GraphSpec:
        document = json.loads(self._canonical_graph)
        return GraphSpec.model_validate(document)


class GraphBuilder:
    """Build one v1alpha1 GraphSpec without inferring IDs, roots, or outputs."""

    def __init__(
        self,
        *,
        metadata: Metadata | JsonObject | object = _MISSING,
        input_schema: JsonObject | object = _MISSING,
        output_schema: JsonObject | object = _MISSING,
        state_schema: JsonObject | object = _MISSING,
        policies: GraphPolicies | JsonObject | object = _MISSING,
    ) -> None:
        self._sealed = False
        for value, path, label in (
            (metadata, "#/metadata", "metadata"),
            (input_schema, "#/inputSchema", "input_schema"),
            (output_schema, "#/outputSchema", "output_schema"),
        ):
            if value is _MISSING:
                raise _fail(
                    BuilderErrorCode.MISSING_REQUIRED,
                    f"required constructor value {label} is missing",
                    path,
                )
        self._metadata = _model_snapshot(metadata, Metadata, "#/metadata")
        self._input_schema = _json_object_snapshot(input_schema, "#/inputSchema")
        self._output_schema = _json_object_snapshot(output_schema, "#/outputSchema")
        self._state_schema = (
            None
            if state_schema is _MISSING
            else _json_object_snapshot(state_schema, "#/stateSchema")
        )
        self._policies = (
            None if policies is _MISSING else _model_snapshot(policies, GraphPolicies, "#/policies")
        )
        self._nodes: list[JsonObject] = []
        self._edges: list[JsonObject] = []
        self._entrypoints: list[str] = []
        self._outputs: dict[str, JsonObject] = {}
        self._node_ids: set[str] = set()
        self._edge_ids: set[str] = set()
        self._entrypoint_ids: set[str] = set()

    def _ensure_open(self, path: str = "#") -> None:
        if self._sealed:
            raise _fail(
                BuilderErrorCode.SEALED,
                "graph builder is sealed after a successful build",
                path,
            )

    def add_node(self, node: NodeSpec | JsonObject) -> GraphBuilder:
        index = len(self._nodes)
        self._ensure_open(f"#/nodes/{index}")
        snapshot = _model_snapshot(node, NodeSpec, f"#/nodes/{index}")
        node_id = cast(str, snapshot["id"])
        if node_id in self._node_ids:
            raise _fail(
                BuilderErrorCode.DUPLICATE_NODE,
                "node ID is already present in this builder",
                f"#/nodes/{index}/id",
            )
        self._node_ids.add(node_id)
        self._nodes.append(snapshot)
        return self

    def add_edge(self, edge: EdgeSpec | JsonObject) -> GraphBuilder:
        index = len(self._edges)
        self._ensure_open(f"#/edges/{index}")
        snapshot = _model_snapshot(edge, EdgeSpec, f"#/edges/{index}")
        edge_id = cast(str, snapshot["id"])
        if edge_id in self._edge_ids:
            raise _fail(
                BuilderErrorCode.DUPLICATE_EDGE,
                "edge ID is already present in this builder",
                f"#/edges/{index}/id",
            )
        self._edge_ids.add(edge_id)
        self._edges.append(snapshot)
        return self

    def add_entrypoint(self, node_id: str) -> GraphBuilder:
        index = len(self._entrypoints)
        self._ensure_open(f"#/entrypoints/{index}")
        snapshot = _string_snapshot(node_id, f"#/entrypoints/{index}", non_empty=True)
        if snapshot in self._entrypoint_ids:
            raise _fail(
                BuilderErrorCode.DUPLICATE_ENTRYPOINT,
                "entrypoint is already present in this builder",
                f"#/entrypoints/{index}",
            )
        self._entrypoint_ids.add(snapshot)
        self._entrypoints.append(snapshot)
        return self

    def add_output(self, name: str, endpoint: Endpoint | JsonObject) -> GraphBuilder:
        attempted_path = "#/outputs"
        if type(name) is str:
            attempted_name = name.replace("~", "~0").replace("/", "~1")
            attempted_path = f"#/outputs/{attempted_name}"
        self._ensure_open(attempted_path)
        output_name = _string_snapshot(name, "#/outputs", non_empty=False)
        path_name = output_name.replace("~", "~0").replace("/", "~1")
        path = f"#/outputs/{path_name}"
        if output_name in self._outputs:
            raise _fail(
                BuilderErrorCode.DUPLICATE_OUTPUT,
                "public output name is already present in this builder",
                path,
            )
        self._outputs[output_name] = _model_snapshot(endpoint, Endpoint, path)
        return self

    def set_policies(self, policies: GraphPolicies | JsonObject) -> GraphBuilder:
        """Replace the complete policy object with an immediate detached snapshot."""

        self._ensure_open("#/policies")
        self._policies = _model_snapshot(policies, GraphPolicies, "#/policies")
        return self

    def enable_strict_typed_ports(self) -> GraphBuilder:
        self._ensure_open("#/policies")
        policies = (
            {}
            if self._policies is None
            else cast(JsonObject, portable_json_snapshot(self._policies))
        )
        if TYPED_PORT_POLICY_KEY in policies:
            raise _fail(
                BuilderErrorCode.INVALID_INPUT,
                "strict typed-port policy is already configured",
                "#/policies/graphengineering.reacher-z.github.io~1typed-ports",
            )
        policies[TYPED_PORT_POLICY_KEY] = strict_typed_ports_policy()
        self._policies = policies
        return self

    def build(self) -> BuiltGraph:
        self._ensure_open()
        if not self._entrypoints:
            raise _fail(
                BuilderErrorCode.MISSING_REQUIRED,
                "at least one explicit entrypoint is required",
                "#/entrypoints",
            )
        if not self._outputs:
            raise _fail(
                BuilderErrorCode.MISSING_REQUIRED,
                "at least one explicit public output is required",
                "#/outputs",
            )

        document: JsonObject = {
            "apiVersion": "graphengineering.reacher-z.github.io/v1alpha1",
            "kind": "Graph",
            "metadata": portable_json_snapshot(self._metadata),
            "inputSchema": portable_json_snapshot(self._input_schema),
            "outputSchema": portable_json_snapshot(self._output_schema),
            "entrypoints": portable_json_snapshot(self._entrypoints),
            "outputs": portable_json_snapshot(self._outputs),
            "nodes": portable_json_snapshot(self._nodes),
            "edges": portable_json_snapshot(self._edges),
        }
        if self._state_schema is not None:
            document["stateSchema"] = portable_json_snapshot(self._state_schema)
        if self._policies is not None:
            document["policies"] = portable_json_snapshot(self._policies)

        compilation = try_compile_graph(document)
        if (
            compilation.graph is None
            or compilation.canonical_graph is None
            or compilation.graph_hash is None
        ):
            raise _fail(
                BuilderErrorCode.CORE_REJECTED,
                "canonical graph compiler rejected the built document",
                "#",
                compilation.diagnostics,
            )
        compiled = compilation.graph

        identity = _create_compiled_graph_identity_from_compiled(compiled)
        built = BuiltGraph(
            _canonical_graph=compilation.canonical_graph,
            graph_hash=compilation.graph_hash,
            identity=identity,
        )
        self._sealed = True
        return built


def graph_builder(
    *,
    metadata: Metadata | JsonObject | object = _MISSING,
    input_schema: JsonObject | object = _MISSING,
    output_schema: JsonObject | object = _MISSING,
    state_schema: JsonObject | object = _MISSING,
    policies: GraphPolicies | JsonObject | object = _MISSING,
) -> GraphBuilder:
    """Create a general builder with an explicit Graph IR envelope."""

    return GraphBuilder(
        metadata=metadata,
        input_schema=input_schema,
        output_schema=output_schema,
        state_schema=state_schema,
        policies=policies,
    )


__all__ = [
    "BuilderErrorCode",
    "BuiltGraph",
    "GraphBuilder",
    "GraphBuilderError",
    "graph_builder",
]

File: packages/core/src/builder.ts
import {
  CanonicalizationError,
  captureCanonicalJson,
  parseFrozenCanonicalJson,
} from "./canonical.js";
import {
  createCompiledGraphIdentityFromCompilation,
  type CompiledGraphIdentity,
} from "./component-identity.js";
import { compileGraph, type CompilerDiagnostic } from "./compiler.js";
import {
  validateEdgeDocument,
  validateEndpointDocument,
  validateMetadataDocument,
  validateNodeDocument,
  validatePoliciesDocument,
} from "./schema-validation.js";
import {
  STRICT_TYPED_PORTS_POLICY,
  STRICT_TYPED_PORTS_POLICY_KEY,
} from "./typed-ports.js";
import type {
  EdgeSpec,
  Endpoint,
  GraphMetadata,
  GraphPolicies,
  GraphSpec,
  JsonSchema,
  NodeSpec,
} from "./types.js";

export type GraphBuilderErrorCode =
  | "GE_BUILDER_INVALID_INPUT"
  | "GE_BUILDER_DUPLICATE_NODE"
  | "GE_BUILDER_DUPLICATE_EDGE"
  | "GE_BUILDER_DUPLICATE_ENTRYPOINT"
  | "GE_BUILDER_DUPLICATE_OUTPUT"
  | "GE_BUILDER_MISSING_REQUIRED"
  | "GE_BUILDER_CORE_REJECTED"
  | "GE_BUILDER_SEALED";

export interface GraphBuilderErrorProjection {
  readonly code: GraphBuilderErrorCode;
  readonly message: string;
  readonly path: string;
  readonly diagnostics: readonly CompilerDiagnostic[];
}

export class GraphBuilderError extends Error implements GraphBuilderErrorProjection {
  readonly code: GraphBuilderErrorCode;
  readonly path: string;
  readonly diagnostics: readonly CompilerDiagnostic[];

  constructor(
    code: GraphBuilderErrorCode,
    message: string,
    path: string,
    diagnostics: readonly CompilerDiagnostic[] = [],
  ) {
    super(message);
    this.name = "GraphBuilderError";
    this.code = code;
    this.path = path;
    const captured = captureCanonicalJson(diagnostics);
    this.diagnostics = parseFrozenCanonicalJson<readonly CompilerDiagnostic[]>(captured.serialized);
  }

  toJSON(): GraphBuilderErrorProjection {
    return Object.freeze({
      code: this.code,
      message: this.message,
      path: this.path,
      diagnostics: this.diagnostics,
    });
  }
}

export interface GraphBuilderOptions {
  readonly metadata: GraphMetadata;
  readonly inputSchema: JsonSchema;
  readonly outputSchema: JsonSchema;
  readonly stateSchema?: JsonSchema;
  readonly policies?: GraphPolicies;
}

export interface BuiltGraph {
  readonly graph: GraphSpec;
  readonly canonicalGraph: string;
  readonly graphHash: string;
  readonly identity: CompiledGraphIdentity;
}

const BUILDER_OPTION_KEYS = new Set([
  "metadata",
  "inputSchema",
  "outputSchema",
  "stateSchema",
  "policies",
]);

function record(value: unknown): value is Readonly<Record<string, unknown>> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function childPath(base: string, capturedPath: string): string {
  if (capturedPath === "#") return base;
  return `${base}${capturedPath.slice(1)}`;
}

function snapshot<T>(value: unknown, path: string): T {
  try {
    return captureCanonicalJson(value).value as unknown as T;
  } catch (error) {
    const failurePath = error instanceof CanonicalizationError
      ? childPath(path, error.path)
      : path;
    throw new GraphBuilderError(
      "GE_BUILDER_INVALID_INPUT",
      "Builder input must be detached portable canonical JSON",
      failurePath,
    );
  }
}

function invalidInput(message: string, path: string): never {
  throw new GraphBuilderError("GE_BUILDER_INVALID_INPUT", message, path);
}

function requireValidFragment(
  issues: readonly string[],
  message: string,
  path: string,
): void {
  if (issues.length > 0) invalidInput(message, path);
}

function outputRecord(entries: readonly (readonly [string, Endpoint])[]): Readonly<Record<string, Endpoint>> {
  const outputs = Object.create(null) as Record<string, Endpoint>;
  for (const [name, endpoint] of entries) {
    Object.defineProperty(outputs, name, {
      value: endpoint,
      enumerable: true,
      configurable: false,
      writable: false,
    });
  }
  return Object.freeze(outputs);
}

/** General, inference-free authoring builder for the current v1alpha1 Graph IR. */
export class GraphBuilder {
  readonly #metadata: GraphMetadata;
  readonly #inputSchema: JsonSchema;
  readonly #outputSchema: JsonSchema;
  readonly #stateSchema: JsonSchema | undefined;
  #policies: GraphPolicies | undefined;
  readonly #nodes: NodeSpec[] = [];
  readonly #edges: EdgeSpec[] = [];
  readonly #entrypoints: string[] = [];
  readonly #outputs: [string, Endpoint][] = [];
  readonly #nodeIds = new Set<string>();
  readonly #edgeIds = new Set<string>();
  readonly #entrypointIds = new Set<string>();
  readonly #outputNames = new Set<string>();
  #sealed = false;

  constructor(options: GraphBuilderOptions | unknown) {
    const captured = snapshot<Readonly<Record<string, unknown>>>(options, "#");
    if (!record(captured)) invalidInput("Builder options must be an object", "#");
    for (const key of Object.keys(captured)) {
      if (!BUILDER_OPTION_KEYS.has(key)) {
        invalidInput(`Unknown builder option '${key}'`, `#/${key}`);
      }
    }
    for (const key of ["metadata", "inputSchema", "outputSchema"] as const) {
      if (!Object.hasOwn(captured, key)) {
        throw new GraphBuilderError(
          "GE_BUILDER_MISSING_REQUIRED",
          `Builder option '${key}' is required`,
          `#/${key}`,
        );
      }
    }
    if (!record(captured.metadata)) invalidInput("metadata must be an object", "#/metadata");
    if (!record(captured.inputSchema)) invalidInput("inputSchema must be an object", "#/inputSchema");
    if (!record(captured.outputSchema)) invalidInput("outputSchema must be an object", "#/outputSchema");
    if (captured.stateSchema !== undefined && !record(captured.stateSchema)) {
      invalidInput("stateSchema must be an object when present", "#/stateSchema");
    }
    if (captured.policies !== undefined && !record(captured.policies)) {
      invalidInput("policies must be an object when present", "#/policies");
    }
    requireValidFragment(
      validateMetadataDocument(captured.metadata),
      "metadata must be a valid Graph IR metadata object",
      "#/metadata",
    );
    if (captured.policies !== undefined) {
      requireValidFragment(
        validatePoliciesDocument(captured.policies),
        "policies must be a valid Graph IR policy object",
        "#/policies",
      );
    }
    this.#metadata = captured.metadata as unknown as GraphMetadata;
    this.#inputSchema = captured.inputSchema as JsonSchema;
    this.#outputSchema = captured.outputSchema as JsonSchema;
    this.#stateSchema = captured.stateSchema as JsonSchema | undefined;
    this.#policies = captured.policies as GraphPolicies | undefined;
  }

  #assertOpen(path = "#"): void {
    if (this.#sealed) {
      throw new GraphBuilderError(
        "GE_BUILDER_SEALED",
        "Graph builder is sealed after build()",
        path,
      );
    }
  }

  addNode(value: NodeSpec | unknown): this {
    this.#assertOpen(`#/nodes/${this.#nodes.length}`);
    const path = `#/nodes/${this.#nodes.length}`;
    const node = snapshot<NodeSpec>(value, path);
    requireValidFragment(
      validateNodeDocument(node, this.#nodes.length),
      "Node must be a valid Graph IR node object",
      path,
    );
    if (this.#nodeIds.has(node.id)) {
      throw new GraphBuilderError(
        "GE_BUILDER_DUPLICATE_NODE",
        `Node id '${node.id}' is already present`,
        `${path}/id`,
      );
    }
    this.#nodeIds.add(node.id);
    this.#nodes.push(node);
    return this;
  }

  addEdge(value: EdgeSpec | unknown): this {
    this.#assertOpen(`#/edges/${this.#edges.length}`);
    const path = `#/edges/${this.#edges.length}`;
    const edge = snapshot<EdgeSpec>(value, path);
    requireValidFragment(
      validateEdgeDocument(edge, this.#edges.length),
      "Edge must be a valid Graph IR edge object",
      path,
    );
    if (this.#edgeIds.has(edge.id)) {
      throw new GraphBuilderError(
        "GE_BUILDER_DUPLICATE_EDGE",
        `Edge id '${edge.id}' is already present`,
        `${path}/id`,
      );
    }
    this.#edgeIds.add(edge.id);
    this.#edges.push(edge);
    return this;
  }

  addEntrypoint(nodeId: string): this {
    const path = `#/entrypoints/${this.#entrypoints.length}`;
    this.#assertOpen(path);
    if (typeof nodeId !== "string" || nodeId.length === 0) {
      invalidInput("Entrypoint must be a non-empty string", path);
    }
    if (this.#entrypointIds.has(nodeId)) {
      throw new GraphBuilderError(
        "GE_BUILDER_DUPLICATE_ENTRYPOINT",
        `Entrypoint '${nodeId}' is already present`,
        path,
      );
    }
    this.#entrypointIds.add(nodeId);
    this.#entrypoints.push(nodeId);
    return this;
  }

  addOutput(name: string, value: Endpoint | unknown): this {
    const path = `#/outputs/${typeof name === "string" ? name.replaceAll("~", "~0").replaceAll("/", "~1") : ""}`;
    this.#assertOpen(path);
    if (typeof name !== "string") invalidInput("Output name must be a string", "#/outputs");
    if (this.#outputNames.has(name)) {
      throw new GraphBuilderError(
        "GE_BUILDER_DUPLICATE_OUTPUT",
        `Output '${name}' is already present`,
        path,
      );
    }
    const endpoint = snapshot<Endpoint>(value, path);
    requireValidFragment(
      validateEndpointDocument(endpoint, path),
      "Output endpoint must be a valid Graph IR endpoint object",
      path,
    );
    this.#outputNames.add(name);
    this.#outputs.push([name, endpoint]);
    return this;
  }

  /** Replace the complete graph policy object with one detached snapshot. */
  setPolicies(value: GraphPolicies | unknown): this {
    this.#assertOpen("#/policies");
    const policies = snapshot<GraphPolicies>(value, "#/policies");
    requireValidFragment(
      validatePoliciesDocument(policies),
      "policies must be a valid Graph IR policy object",
      "#/policies",
    );
    this.#policies = policies;
    return this;
  }

  enableStrictTypedPorts(): this {
    this.#assertOpen("#/policies");
    if (
      this.#policies !== undefined &&
      Object.hasOwn(this.#policies, STRICT_TYPED_PORTS_POLICY_KEY)
    ) {
      invalidInput(
        "Strict typed-port policy is already configured and will not be overwritten",
        "#/policies/graphengineering.reacher-z.github.io~1typed-ports",
      );
    }
    const policies = Object.create(null) as Record<string, unknown>;
    for (const [key, value] of Object.entries(this.#policies ?? {})) {
      Object.defineProperty(policies, key, {
        value,
        enumerable: true,
        configurable: true,
        writable: true,
      });
    }
    Object.defineProperty(policies, STRICT_TYPED_PORTS_POLICY_KEY, {
      value: STRICT_TYPED_PORTS_POLICY,
      enumerable: true,
      configurable: true,
      writable: true,
    });
    this.#policies = snapshot<GraphPolicies>(policies, "#/policies");
    return this;
  }

  build(): BuiltGraph {
    this.#assertOpen("#");
    this.#sealed = true;
    if (this.#entrypoints.length === 0) {
      throw new GraphBuilderError(
        "GE_BUILDER_MISSING_REQUIRED",
        "At least one explicit entrypoint is required",
        "#/entrypoints",
      );
    }
    if (this.#outputs.length === 0) {
      throw new GraphBuilderError(
        "GE_BUILDER_MISSING_REQUIRED",
        "At least one explicit public output is required",
        "#/outputs",
      );
    }

    const candidate: Record<string, unknown> = {
      apiVersion: "graphengineering.reacher-z.github.io/v1alpha1",
      kind: "Graph",
      metadata: this.#metadata,
      inputSchema: this.#inputSchema,
      outputSchema: this.#outputSchema,
      entrypoints: [...this.#entrypoints],
      outputs: outputRecord(this.#outputs),
      nodes: [...this.#nodes],
      edges: [...this.#edges],
    };
    if (this.#stateSchema !== undefined) candidate.stateSchema = this.#stateSchema;
    if (this.#policies !== undefined) candidate.policies = this.#policies;

    const compilation = compileGraph(candidate);
    if (!compilation.valid || compilation.canonicalGraph === null || compilation.graphHash === null) {
      throw new GraphBuilderError(
        "GE_BUILDER_CORE_REJECTED",
        "Core compiler rejected the constructed graph",
        "#",
        compilation.diagnostics,
      );
    }
    const graph = parseFrozenCanonicalJson<GraphSpec>(compilation.canonicalGraph);
    return Object.freeze({
      graph,
      canonicalGraph: compilation.canonicalGraph,
      graphHash: compilation.graphHash,
      identity: createCompiledGraphIdentityFromCompilation(compilation),
    });
  }
}

export function graphBuilder(options: GraphBuilderOptions | unknown): GraphBuilder {
  return new GraphBuilder(options);
}
